src/display/optional_content_config.js

```diff
@@ -0,0 +1,107 @@
+/* Copyright 2020 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { assert, warn } from "../shared/util.js";
+
+class OptionalContentGroup {
+  constructor(name, intent) {
+    this.visible = true;
+    this.name = name;
+    this.intent = intent;
+  }
+}
+
+class OptionalContentConfig {
+  constructor(data) {
+    this.name = null;
+    this.creator = null;
+    this.groups = new Map();
+
+    if (data === null) {
+      return;
+    }
+    this.name = data.name;
+    this.creator = data.creator;
+    for (const group of data.groups) {
+      this.groups.set(
+        group.id,
+        new OptionalContentGroup(group.name, group.intent)
+      );
+    }
+
+    if (data.baseState === "OFF") {
+      for (const group of this.groups) {
+        group.visible = false;
+      }
+    }
+
+    for (const on of data.on) {
+      this.groups.get(on).visible = true;
+    }
+
+    for (const off of data.off) {
+      this.groups.get(off).visible = false;
+    }
+  }
+
+  isVisible(group) {
```

Comment:
For PDF documents where the `OptionalContentConfig` is initialized with `data === null`, such as `issue7769` in the test-suite, where `this.groups` is thus *always empty* we still need to ensure that rendering works correctly. Hence I'd suggest, in addition to the inline checks mentioned elsewhere, that you also add the following checks here:
```suggestion
  isVisible(group) {
    if (!this.groups.size) {
      warn("No optional content groups found.");
      return true;
    }
```

Reply:
Section 8.11.4.2 in the standard says this about the `OCProperties` entry which is missing in issue7769:
>if it is missing, a conforming reader shall 
ignore any optional content structures in the document.
I think this means that all graphics should be made visible.